x-pack/plugins/observability_solution/observability_ai_assistant/server/service/knowledge_base_service/recall_from_search_connectors.ts

```diff
@@ -0,0 +1,284 @@
+
+  const params = {
+    index: connectorIndices,
+    size: 20,
+    _source: {
+      excludes: semanticTextFields.map((field) => `${field}.inference`),
+    },
+    query: {
+      bool: {
+        should: semanticTextFields.flatMap((field) => {
+          return queries.map(({ text, boost = 1 }) => ({
+            bool: { filter: [{ semantic: { field, query: text, boost } }] },
+          }));
+        }),
+        minimum_should_match: 1,
+      },
+    },
+  };
+
+  const response = await esClient.asCurrentUser.search<unknown>(params);
+
+  const results = response.hits.hits.map((hit) => ({
+    text: JSON.stringify(hit._source),
+    score: hit._score!,
+    is_correction: false,
+    id: hit._id!,
+  }));
+
+  return results;
+}
+
+async function recallFromLegacyConnectors({
+  queries,
+  esClient,
+  logger,
+  core,
+  connectorIndices,
+}: {
+  queries: Array<{ text: string; boost?: number }>;
+  esClient: { asCurrentUser: ElasticsearchClient; asInternalUser: ElasticsearchClient };
+  uiSettingsClient: IUiSettingsClient;
+  logger: Logger;
+  core: CoreSetup<ObservabilityAIAssistantPluginStartDependencies>;
+  connectorIndices: string[] | undefined;
+}): Promise<RecalledEntry[]> {
+  const ML_INFERENCE_PREFIX = 'ml.inference.';
+
+  const fieldCaps = await esClient.asCurrentUser.fieldCaps({
+    index: connectorIndices,
+    fields: `${ML_INFERENCE_PREFIX}*`,
+    allow_no_indices: true,
+    types: ['sparse_vector'],
+    filters: '-metadata,-parent',
+  });
+
+  const fieldsWithVectors = Object.keys(fieldCaps.fields).map((field) =>
+    field.replace('_expanded.predicted_value', '').replace(ML_INFERENCE_PREFIX, '')
+  );
+
+  if (!fieldsWithVectors.length) {
+    return [];
+  }
+
+  const modelId = await getElserModelId(core, logger);
+  const esQueries = fieldsWithVectors.flatMap((field) => {
+    const vectorField = `${ML_INFERENCE_PREFIX}${field}_expanded.predicted_value`;
+    const modelField = `${ML_INFERENCE_PREFIX}${field}_expanded.model_id`;
+
+    return queries.map(({ text, boost = 1 }) => {
+      return {
+        bool: {
+          should: [
+            {
+              text_expansion: {
+                [vectorField]: {
+                  model_text: text,
+                  model_id: modelId,
+                  boost,
+                },
+              },
+            },
+          ],
+          filter: [
+            {
+              term: {
+                [modelField]: modelId,
+              },
+            },
+          ],
+        },
+      };
+    });
+  });
+
+  const response = await esClient.asCurrentUser.search<unknown>({
+    index: connectorIndices,
+    size: 20,
+    _source: {
+      exclude: ['_*', 'ml*'],
+    },
+    query: {
+      bool: {
+        should: esQueries,
+      },
+    },
+  });
+
+  const results = response.hits.hits.map((hit) => ({
+    text: JSON.stringify(hit._source),
+    score: hit._score!,
+    is_correction: false,
+    id: hit._id!,
+  }));
+
+  return results;
+}
```

Comment:
FYI: This is not new code. This is what we have already for legacy (aka non semantic text) connectors